### PR TITLE
iox-#1394 iox-#1196 Address Axivion and clang-tidy findings for `cxx::function_ref`

### PR DESCRIPTION
**Pull request by mossmaurice**
Signed-off-by: Simon Hoinkis <simon.hoinkis@apex.ai>

## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] ~~Changelog updated [in the unreleased section][changelog] including API breaking changes~~
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer
* Address `AUTOSAR` warnings with Axivion
* Fix clang-tidy warnings in `cxx::function_ref`
* Findings regardings `Expects` not considered will be addressed as part of #1032 by @MatthiasKillat 

### Todo
* [x] Some suppressions not considered by Axivion -> Fixed
* [x] ~~`M7-1-2`: Shall constness of 'target' be changed?~~
* [x] Discuss `A2-7-3` and `A14-5-2`

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1394 (partly)
- Closes #1196 (partly)

### CODE REVIEW

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 1]
```diff
@@ -34,12 +49,17 @@ template <class ReturnType, class... ArgTypes>
...
+    // AXIVION Next Line AutosarC++19_03-A5.2.4, AutosarC++19_03-A5.2.4-M5.2.6 : Type-safety ensured by casting back on
+    // call
     m_pointerToCallable = reinterpret_cast<void*>(function);
 
     // the lambda does not capture and is thus convertible to a function pointer
     // (required by the C++ standard)
     m_functionPointer = [](void* target, ArgTypes... args) -> ReturnType {
-        auto f = reinterpret_cast<ReturnType (*)(ArgTypes...)>(target);
+        using PointerType = ReturnType (*)(ArgTypes...);
+        // AXIVION Next Line AutosarC++19_03-A5.2.4 : Type-safety ensured by casting from type
```
Comment (mossmaurice) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:60]: @saif-at-github Axivion is having problems to detect the suppression of `AutosarC++19_03-A5.2.4`. I still get 4 hits when analyzing this branch? Do you know why this is happening? @dkroenke had the same issue.

**Review by dkroenke — DISMISSED** [review]

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp [hunk 2]
```diff
@@ -61,41 +67,37 @@ class function_ref;
...
-        bool,
-        bool(std::is_same<typename std::decay<T1>::type, typename std::decay<T2>::type>::value)>;
-
   public:
     ~function_ref() noexcept = default;
 
     function_ref(const function_ref&) noexcept = default;
 
-    function_ref& operator=(const function_ref&) noexcept = default;
+    function_ref& operator=(const function_ref&) & noexcept = default;
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp:77]: Why did you add a reference here? Was it a warning? Should we do this for all copy assignments?

So this rule forbids code like:
```cxx
std::move(a) = b;
```
Which is valid and well defined but on the first glimpse weird. But take a look at this here

```cxx
template<typename T>
void initializeSomeArgs(T&& args) {
    std::forward<T>(args) = someT;
}
```

Here we take `args` as universal reference (so sometimes also as rvalue reference) and would like to assign a new value to it. 
Why should we forbid that? What kind of error can possible originate from that?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 3]
```diff
@@ -15,15 +15,34 @@
...
+
+#include <memory>
+
 namespace iox
 {
 namespace cxx
 {
 template <class ReturnType, class... ArgTypes>
 template <typename CallableType, typename>
+// AXIVION Next Line AutosarC++19_03-A12.1.2 : Members are initialized in the same manner, NSDMI with nullptr is
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:32]: In what manner are the members initialized? What does NSDMI mean?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 4]
```diff
@@ -15,15 +15,34 @@
...
+// explicit
+// AXIVION Next Line AutosarC++19_03-A8.4.6 : Only ArgTypes needs to be forwarded
 inline function_ref<ReturnType(ArgTypes...)>::function_ref(CallableType&& callable) noexcept
+    // AXIVION Next Construct AutosarC++19_03-A5.2.4, AutosarC++19_03-A5.2.3, CertC++-EXP55 : Type-safety ensured by
+    // casting back on call
+    // NOLINTNEXTLINE (cppcoreguidelines-pro-type-reinterpret-cast, cppcoreguidelines-pro-type-const-cast)
     : m_pointerToCallable(const_cast<void*>(reinterpret_cast<const void*>(std::addressof(callable))))
+    // AXIVION Next Line AutosarC++19_03-A15.4.4 : Lambda not 'noexcept' as callable might throw
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:40]: With this argument the call operator must be `noexcept`.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 5]
```diff
@@ -81,10 +107,13 @@ inline void function_ref<ReturnType(ArgTypes...)>::swap(function_ref<ReturnType(
 }
 
 template <class ReturnType, class... ArgTypes>
+// AXIVION Next Line AutosarC++19_03-A2.10.4 : Overload for swap(function_ref, function_ref) as in STL
 inline void swap(function_ref<ReturnType(ArgTypes...)>& lhs, function_ref<ReturnType(ArgTypes...)>& rhs) noexcept
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:111]: Could you please declare this in the header. This is public API but declared and defined hidden in the `inl` file. But the include of the inl file at the end of the header makes this again public.

And as you stated this is also defined in the STL so it absolutely makes sense to provide the same functionality correctly to the user.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 6]
```diff
@@ -32,16 +51,22 @@ inline function_ref<ReturnType(ArgTypes...)>::function_ref(CallableType&& callab
...
+    // back on call NOLINTNEXTLINE (cppcoreguidelines-pro-type-reinterpret-cast)
+    : m_pointerToCallable(reinterpret_cast<void*>(function))
+    ,
     // the lambda does not capture and is thus convertible to a function pointer
     // (required by the C++ standard)
-    m_functionPointer = [](void* target, ArgTypes... args) -> ReturnType {
-        auto f = reinterpret_cast<ReturnType (*)(ArgTypes...)>(target);
+    m_functionPointer([](void* target, ArgTypes... args) -> ReturnType {
+        using PointerType = ReturnType (*)(ArgTypes...);
+        // AXIVION Next Construct AutosarC++19_03-A5.2.4 : Type-safety ensured by casting from type
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:63]: Type safety is ensured by the fact we only pass functions with this type as `target`, i.e. they are convertible.

This is entirely under the control of the class.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 7]
```diff
@@ -52,7 +77,7 @@ inline function_ref<ReturnType(ArgTypes...)>::function_ref(function_ref&& rhs) n
 
 template <class ReturnType, class... ArgTypes>
 inline function_ref<ReturnType(ArgTypes...)>&
-function_ref<ReturnType(ArgTypes...)>::operator=(function_ref<ReturnType(ArgTypes...)>&& rhs) noexcept
+function_ref<ReturnType(ArgTypes...)>::operator=(function_ref<ReturnType(ArgTypes...)>&& rhs) & noexcept
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:80]: lvalue restriction is not necessary.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 8]
```diff
@@ -69,6 +94,7 @@ template <class ReturnType, class... ArgTypes>
 inline ReturnType function_ref<ReturnType(ArgTypes...)>::operator()(ArgTypes... args) const noexcept
 {
     // Expect that a callable was assigned beforehand
+    // AXIVION Next Line AutosarC++19_03-M5.3.1 : 'nullptr' check shall be performed explicitly
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:97]: Could this be avoided by casting the string literal to bool instead? It is of not much importance as "Ensures" will be removed in the future.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 9]
```diff
@@ -81,10 +107,13 @@ inline void function_ref<ReturnType(ArgTypes...)>::swap(function_ref<ReturnType(
 }
 
 template <class ReturnType, class... ArgTypes>
+// AXIVION Next Line AutosarC++19_03-A2.10.4 : Overload for swap(function_ref, function_ref) as in STL
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:110]: Why is this needed at all? If this rule is about not using STL names it is problematic for us as we reimplement some of the STL functionality and keep the function names like in STL where possible.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp:77]: @elfenpiff That's due to AUTOSAR  rule `A12-8-7`. Indeed

```
std::move(a) = b;
```
is weird and does not make sense. You can find more example in the AUTOSAR document.

For your 2nd example, I suppose if it is needed in some cases, we should explicitly define `foo& operator=(const foo&) &&` and if desired call `operator=() &` from there explicitly.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl (hunk 3 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl]: @elfenpiff https://www.giybf.com :upside_down_face:

@MatthiasKillat Correct, I switchted it to `Next Construct`.

The rule just says either init via the `: m_foo(foo)` or inline via `Type m_foo{defaultValue}`, but don't mix things as you might forget it in another c'tor. However, doing it explicitly in both ways, is the best IMHO.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl (hunk 4 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:40]: @elfenpiff I don't get this, the `operator()()` is already `noexcept`.

@MatthiasKillat Correct, we need to monitor that.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl (hunk 5 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:112]: Sure, fixed!

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp [hunk 10]
```diff
@@ -15,15 +15,15 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
+// AXIVION DISABLE STYLE AutosarC++19_03-A16.2.3 : <type_traits> is included
+
 #ifndef IOX_HOOFS_CXX_FUNCTION_REF_HPP
 #define IOX_HOOFS_CXX_FUNCTION_REF_HPP
 
+// AXIVION Next Line AutosarC++19_03-A16.2.2 : Needed for Expects and Ensures macros
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp:23]: AFAIK Axivion does not detect that the macro from `requires.hpp` is used.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 11]
```diff
@@ -15,15 +15,34 @@
...
+// AXIVION Next Line AutosarC++19_03-A8.4.6 : Only ArgTypes needs to be forwarded
 inline function_ref<ReturnType(ArgTypes...)>::function_ref(CallableType&& callable) noexcept
+    // AXIVION Next Construct AutosarC++19_03-A5.2.4, AutosarC++19_03-A5.2.3, CertC++-EXP55 : Type-safety ensured by
+    // casting back on call
+    // NOLINTNEXTLINE (cppcoreguidelines-pro-type-reinterpret-cast, cppcoreguidelines-pro-type-const-cast)
     : m_pointerToCallable(const_cast<void*>(reinterpret_cast<const void*>(std::addressof(callable))))
+    // AXIVION Next Line AutosarC++19_03-A15.4.4 : Lambda not 'noexcept' as callable might throw
     , m_functionPointer([](void* target, ArgTypes... args) -> ReturnType {
+        // AXIVION Next Construct AutosarC++19_03-A5.2.4, CertC++-EXP36 : Type-safety ensured by casting from type
+        // AXIVION Next Construct AutosarC++19_03-A5.3.2, AutosarC++19_03-M5.2.8 : Check for 'nullptr' is performed on
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:43]: @MatthiasKillat True, we'll have to closely monitor that in future.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl (hunk 7 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:81]: See your comment in the other thread. The restriction makes sense and `operator=() &&` should be explicitly defined if ever needed.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl (hunk 8 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:98]: @MatthiasKillat Jep, see description of this PR. This topic is left open for you to address.

**Comment by mossmaurice** [comment]
@elfenpiff @MatthiasKillat @elBoberido 

The combination of `clang-tidy` and Axivion suppression has been confirmed to work in a call with Axivion yesterday:

```
// AXIVION Next Construct Ruleset-A1.2.3 : Because foo and bar
// NOLINTNEXTLINE (foobar)
```

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp:23]: -1 for the tool I guess

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp (hunk 2 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp:77]: If both implementations (`&&` and `&`) exist and are identical it does not make sense to distinguish. Only if one should not exist (as here) or if they both exist but are different.

File: iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp [hunk 12]
```diff
@@ -61,41 +67,37 @@ class function_ref;
...
     /// @param[in] callable that is not a function_ref
     template <typename CallableType,
-              typename = std::enable_if_t<!is_function_pointer<CallableType>::value
-                                          && !has_same_decayed_type<CallableType, function_ref>::value
-                                          && is_invocable<CallableType, ArgTypes...>::value>>
-    function_ref(CallableType&& callable) noexcept;
+              typename = std::enable_if_t<(!is_function_pointer<CallableType>::value)
+                                          && (!has_same_decayed_type<CallableType, function_ref>::value)
+                                          && (is_invocable<CallableType, ArgTypes...>::value)>>
+    // AXIVION Next Line AutosarC++19_03-A12.1.4 : Implicit conversion is needed for lambdas
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/cxx/function_ref.hpp:85]: I meant that more accurately it is needed to pass any kind of functors conveniently to functions with `function_ref` arguments. It is not impossible to use lambdas or functors with an explicit version. If it is explicit it just would be inconvenient as the constructor of `function_ref` would have to be explicitly called.

`required to pass lambdas to interfaces with function_ref arguments` would be more accurate.

By the way that is a similar reason as to why the relative pointers work as they do, to avoid inconvenience while passing raw pointers to interfaces or variables that take relative pointers (or generally fancy pointers). This is basically the posterchild example why there need to be exceptions to this explicit ctor rule.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 13]
```diff
@@ -15,15 +15,34 @@
...
+// explicit
+// AXIVION Next Line AutosarC++19_03-A8.4.6 : Only ArgTypes needs to be forwarded
 inline function_ref<ReturnType(ArgTypes...)>::function_ref(CallableType&& callable) noexcept
+    // AXIVION Next Construct AutosarC++19_03-A5.2.4, AutosarC++19_03-A5.2.3, CertC++-EXP55 : Type-safety ensured by
+    // casting back on call
+    // NOLINTNEXTLINE (cppcoreguidelines-pro-type-reinterpret-cast, cppcoreguidelines-pro-type-const-cast)
     : m_pointerToCallable(const_cast<void*>(reinterpret_cast<const void*>(std::addressof(callable))))
+    // AXIVION Next Line AutosarC++19_03-A15.4.4 : Lambda not 'noexcept' as callable might throw
     , m_functionPointer([](void* target, ArgTypes... args) -> ReturnType {
+        // AXIVION Next Construct AutosarC++19_03-A5.2.4, CertC++-EXP36 : Type-safety ensured by casting from type
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:42]: @mossmaurice I still think the justification is not accurate/understandable. type erasure should be mentioned.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl [hunk 14]
```diff
@@ -15,15 +15,34 @@
...
+        // AXIVION Next Construct AutosarC++19_03-A5.3.2, AutosarC++19_03-M5.2.8 : Check for 'nullptr' is performed on
+        // call
+        // NOLINTNEXTLINE (cppcoreguidelines-pro-type-reinterpret-cast)
```
[the remark this answers is not in the document]
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:45]: OK, I am still not sure about the rules when both tools are used at the same time (not decided to my knowledge and not for this PR).

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl (hunk 6 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:64]: @mossmaurice I still find the justification lacking here, hence my comment about the reason.

Without extra knowledge of how it works, I could not infer it from the justification.

More precisely
`the class design ensures a cast to the actual type of target`

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl (hunk 9 above)
Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/function_ref.inl:111]: It is probably not a good rule for us to activate then ... but also not a question for this PR

**Review by MatthiasKillat — APPROVED** [review]

**Review by dkroenke — APPROVED** [review]
